x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java

```diff
@@ -244,4 +261,27 @@ public static boolean allocationAllowed(String tierName, Set<DiscoveryNodeRole>
         }
         return false;
     }
+
+    /**
+     * Creates a new DiscoveryNodes object representing the state of the cluster without nodes that are shutting down for removal.
+     * @param nodes the nodes currently present in the cluster
+     * @param nodesShutdownMetadata shutdown metadata, so we know which nodes to filter
+     * @return the nodes in the cluster which are not currently marked for removal
+     */
+    private static DiscoveryNodes filterRemovingNodes(DiscoveryNodes nodes, NodesShutdownMetadata nodesShutdownMetadata) {
+        List<String> removingNodeIds = nodesShutdownMetadata.getAll()
```

Comment:
Looks like this is mostly going to be empty, can we avoid allocating all the stream stuff on the hot path here please? Also, ++ to Ievgen, we shouldn't have to build a new `DiscoveryNodes` here.